### ISSUE TITLE
poor diagnostic on invalid format string inside a lambda

### ISSUE THREAD

**Issue by janwilmans** [llvm/llvm-project#87406]
I know I'm using {fmt} here in the example but that is for the sake of comparison, the diagnostic is similarly bad with std::format, see the second godbolt link at the botton for an {fmt}-free example.

```
#include <fmt/core.h>
#include <string>
#include <functional>

// std::string example()
// {
//     return fmt::format("{.3}\n", 0);
// }

std::function<std::string()> example = []
{
    return fmt::format("{.3}\n", 0);
};

int main()
{
    return example().size();
}
```

when the format-string is invalid inside the lamdba, the diagnostic is very confusing. I can understand why, and I'm sure anything can be done about it, however, it seems to me that the error-handling strategy used in {fmt} is not working anymore in later versions of clang.

[godbolt link](https://cppcoach.godbolt.org/z/crnevsWeK) using {fmt} as an example of the `throw_format_error("invalid format string");` being displayed.
[godbolt link](https://cppcoach.godbolt.org/z/qTYf7accE) using std::format showing that inside a lamdba the std::format diagnostic is very poor

**Comment by shafik** [comment]
CC @mordante @ldionne

**Comment by mordante** [comment]
Based on the first example for `{fmt}` this seems a regression between clang-16 and clang-17.
I tested with clang and libc++ instead of libstdc++ https://cppcoach.godbolt.org/z/6K9s7j6x8 and get similar results as `{fmt}`. So I expect it's a Clang regression.

**Comment by janwilmans** [comment]
I suspect the std::format implementation uses a "call to non-constexpr function" to deliberately stop further evaluation and basically trick the diagnostic into emitting a relevant error message.

In the happy flow, that path is not evaluated and the error is not detected.
However in later compilers that "call to non-constexpr function" itself is no longer an error, the evaluation continues to trash on, goes off into the deep end and comes back with a rather obscure message.

**Comment by mordante** [comment]
> I suspect the std::format implementation uses a "call to non-constexpr function" to deliberately stop further evaluation and basically trick the diagnostic into emitting a relevant error message.

Yes, the other option would be to mark `__throw_format_error` `constexpr` and use the `throw` to make the format string evaluation not a core constant expression.
